# Mandatory Exercise - Session 6
## Students: Nafis Banirazi & Jan Carbonell
### Lab Objective:
The Objective of this lab are the following:

To read all pairs of sentences from a trial set, compute their similarities considering Lesk's algorithm and using the Jaccard distance.
Compare the results with those on session 2 and 3.
Compare the results with the gold standard by giving the Pearson correlation.
Answering if words or lemmas perfom better and if we could extrapolate that to any pair of texts.

In [1]:
#intial set of imports
import nltk
from nltk import pos_tag
from nltk.wsd import lesk
from nltk.metrics import jaccard_distance
from nltk.corpus import wordnet as wn
from scipy.stats import pearsonr

#variable initialization and instantiation
d = {}
tests = []
corpus = []
tests = []
standard = []

## 1. Read all pairs of sentences of the trial set within the evaluation framework of the project.
We proceed to open and read the input file

In [2]:
for line in open('./trial/STS.input.txt','r'):
    line = line.strip().split("\t")
    if line[0] in d:
        print('wtf')
        pass
    else:
        d[line[0]] = [line[1].lower(), line[2].lower()] #.lower(),line[2].lower()]
    
    
for key in d:
    print(d[key])
print(d)
d3 = dict(d)

['the bird is bathing in the sink.', 'birdie is washing itself in the water basin.']
['in may 2010, the troops attempted to invade kabul.', 'the us army invaded kabul on may 7th last year, 2010.']
['john said he is considered a witness but not a suspect.', '"he is not a suspect anymore." john said.']
['they flew out of the nest in groups.', 'they flew into the nest together.']
['the woman is playing the violin.', 'the young lady enjoys listening to the guitar.']
['john went horse back riding at dawn with a whole group of friends.', 'sunrise at dawn is a magnificent view to take in if you wake up early enough for it.']
{'id1': ['the bird is bathing in the sink.', 'birdie is washing itself in the water basin.'], 'id2': ['in may 2010, the troops attempted to invade kabul.', 'the us army invaded kabul on may 7th last year, 2010.'], 'id3': ['john said he is considered a witness but not a suspect.', '"he is not a suspect anymore." john said.'], 'id4': ['they flew out of the nest in groups.',

## 2. Apply Lesk’s algorithm to the words in the sentences.
In this step, we must first tokenize the sentences. 
We then apply pos_tag to determine the type of each word in the sentence; and apply lesk algorithm.
Then we take the sets to unique values and delete None element.
Finally, we compute the jaccard similarity as 1 - jaccard distance.

In [3]:
def lesking_sentence(pos_tagged_sentence):
    """
    Returns a sentence as the given sentece using lesker algorithms.
    The input sentence must be a pos_tagged sentence (e.g. [('The', 'DN'),
    ('sun', 'NN')]).
    """
    sentence = [i[0] for i in pos_tagged_sentence]
    result = []
    
    none_type_objects = []
    for word, tag in pos_tagged_sentence:
        # 'NoneType' object has no attribute 'name'
        try:
            if tag.startswith('N'):
                result.append(lesk(sentence, word, wn.NOUN).name())
            #if word is a verb
            elif tag.startswith('V'):
                result.append(lesk(sentence, word, wn.VERB).name())
            #if word is an adjective
            elif tag.startswith('J'):
                result.append(lesk(sentence, word, wn.ADJ).name())
            #if word is a verb
            elif tag.startswith('R'):
                result.append(lesk(sentence, word, wn.ADV).name())
            else:
                result.append(word)
        except:
            print('EXEPTION ERROR:', word, tag)
            none_type_objects.append([word, tag, lesk(sentence, word, wn.NOUN)])
            
    return result

In [4]:
for key in d:
    
    # Tokenizing
    sents = [nltk.word_tokenize(s) for s in d[key]];    
    
    w1 = sents[0]
    w2 = sents[1]

    # Determining the category of words
    wordtype1 = pos_tag(w1)    
    wordtype2 = pos_tag(w2)

    # As Lesk's algorithm works with synsets
    # Not every word will be accepted. Just nouns, verbs, adjectives and adverbs
    #determining the definition of words
    lesk1 = lesking_sentence(wordtype1)
    print('Post_Lesked_Sentence: ', lesk1, '\n')
    lesk2 = lesking_sentence(wordtype2)
    print('Post_Lesked_Sentence: ', lesk2, '\n')

    # Removing duplicates
    set1 = set(lesk1)
    set2 = set(lesk2)
    
    #computing jaccard similarity
    jaccard_similarity = 1 - jaccard_distance(set1, set2)
    tests.append(round(jaccard_similarity,3))
print(tests)

Post_Lesked_Sentence:  ['the', 'bird.n.02', 'be.v.12', 'bathe.v.01', 'in', 'the', 'sinkhole.n.01', '.'] 

Post_Lesked_Sentence:  ['shuttlecock.n.01', 'be.v.12', 'wash.v.09', 'itself', 'in', 'the', 'body_of_water.n.01', 'washbasin.n.01', '.'] 

Post_Lesked_Sentence:  ['in', 'may', '2010', ',', 'the', 'troop.n.02', 'undertake.v.01', 'to', 'invade.v.04', 'kabul.n.01', '.'] 

EXEPTION ERROR: army VBP
EXEPTION ERROR: invaded JJ
Post_Lesked_Sentence:  ['the', 'us', 'kabul.n.01', 'on', 'may', '7th', 'last.a.02', 'year.n.03', ',', '2010', '.'] 

Post_Lesked_Sentence:  ['whoremaster.n.01', 'suppose.v.01', 'he', 'embody.v.02', 'view.v.02', 'a', 'witness.n.05', 'but', 'not.r.01', 'a', 'defendant.n.01', '.'] 

Post_Lesked_Sentence:  ['``', 'he', 'embody.v.02', 'not.r.01', 'a', 'defendant.n.01', 'anymore.r.01', '.', "''", 'whoremaster.n.01', 'suppose.v.01', '.'] 

EXEPTION ERROR: nest JJS
Post_Lesked_Sentence:  ['they', 'fly.v.12', 'out', 'of', 'the', 'in', 'group.n.02', '.'] 

EXEPTION ERROR: nest

## 3. Compute their similarities via the Jaccard coefficient.

done previously within the loop

## 4. Compare the results with those in session 2 (document) and 3 (morphology) in which words and lemmas were considered.
The gold standard results would be 5,4,3,2,1,0.


In session 2 we got the following results:

* *'id1': 0.308* --> 1.54
* *'id2': 0.263* --> 1.31
* *'id3': 0.467* --> 2.34
* *'id4': 0.455* --> 2.27
* *'id5': 0.231* --> 1.15
* *'id6': 0.138* --> 0.69

In session 3, using lemmas similarity, we obtained:
* *'id1': 0.333* --> 1.67
* *'id2': 0.412* --> 2.06
* *'id3': 0.571* --> 2.86
* *'id4': 0.333* --> 1.67
* *'id5': 0.167* --> 0.84
* *'id6': 0.138* --> 0.69

And our current result is:
* *'id1': 0.333* --> 1.67
* *'id2': 0.375* --> 1.88
* *'id3': 0.571* --> 2.86
* *'id4': 0.273* --> 1.36
* *'id5': 0.167* --> 0.84
* *'id6': 0.103* --> 0.52

## 5. Compare the results with gold standard by giving the pearson correlation between them.
And now, we open the Golden Standard file and calculate the perason correlation with lesk.

In [5]:
# CAREFUL WITH THE STS.gs FILE. 
# NEEDS TO BE CORRECTED TO 54321 instead of 12345

for line in open('./trial/STS.gs.txt','r'):
    line = line.strip().split("\t")
    standard.append(int(line[1]))
    
a = pearsonr(standard[0:6], tests[0:6])[0]
print('Pearson correlation with lesk:', round(a,3))

Pearson correlation with lesk: 0.668


We compare the results with those in session 2 (document) and 3 (morphology) in which words and lemmas were considered:
- Pearson correlation without lemmatization: 0.396
- Pearson correlation with lemmatization: 0.578
- Pearson correlation with lesk: 0.668

This may also be to an improvement in our way of splitting the words and sorting the sentences that we have implemented in this lab. 



## Optional Exercise
### Stopwords


{'id1': ['the bird is bathing in the sink.', 'birdie is washing itself in the water basin.'], 'id2': ['in may 2010, the troops attempted to invade kabul.', 'the us army invaded kabul on may 7th last year, 2010.'], 'id3': ['john said he is considered a witness but not a suspect.', '"he is not a suspect anymore." john said.'], 'id4': ['they flew out of the nest in groups.', 'they flew into the nest together.'], 'id5': ['the woman is playing the violin.', 'the young lady enjoys listening to the guitar.'], 'id6': ['john went horse back riding at dawn with a whole group of friends.', 'sunrise at dawn is a magnificent view to take in if you wake up early enough for it.']}

## We apply lesk and compute the jaccard distance and its corresponding pearson coefficient:

In [7]:
from nltk.corpus import stopwords

without_stop_words_tests = []

def words_from_sent(sent):
    # tokenized
    tokenized = nltk.word_tokenize(sent)
    # remove stopwords and return
    return [word for word in tokenized if word not in nltk.corpus.stopwords.words('english')]

sw=set(stopwords.words('english'))
for key in d3:
    
    # Tokenizing
    #sent = words_from_sent(d3[key]);    
    
    
    w1 = words_from_sent(d3[key][0])
    w2 = words_from_sent(d3[key][1])

    # Determining the category of words
    wordtype1 = pos_tag(w1)    
    wordtype2 = pos_tag(w2)

    # As Lesk's algorithm works with synsets
    # Not every word will be accepted. Just nouns, verbs, adjectives and adverbs
    #determining the definition of words
    lesk1 = lesking_sentence(wordtype1)
    print('Post_Lesked_Sentence: ', lesk1, '\n')
    lesk2 = lesking_sentence(wordtype2)
    print('Post_Lesked_Sentence: ', lesk2, '\n')

    # Removing duplicates
    set1 = set(lesk1)
    set2 = set(lesk2)
    
    #computing jaccard similarity
    jaccard_similarity = 1 - jaccard_distance(set1, set2)
    without_stop_words_tests.append(round(jaccard_similarity,3))
print(tests)

Post_Lesked_Sentence:  ['bird.n.02', 'bathe.v.03', 'sinkhole.n.01', '.'] 

Post_Lesked_Sentence:  ['shuttlecock.n.01', 'wash.v.09', 'water_system.n.02', 'washbasin.n.01', '.'] 

EXEPTION ERROR: invade JJ
Post_Lesked_Sentence:  ['may', '2010', ',', 'troop.n.04', 'undertake.v.01', 'kabul.n.01', '.'] 

EXEPTION ERROR: army VBP
EXEPTION ERROR: invaded JJ
Post_Lesked_Sentence:  ['us', 'kabul.n.01', 'may', '7th', 'last.a.02', 'year.n.02', ',', '2010', '.'] 

EXEPTION ERROR: witness JJ
Post_Lesked_Sentence:  ['john.n.03', 'suppose.v.01', 'view.v.02', 'suspect.n.01', '.'] 

Post_Lesked_Sentence:  ['``', 'suspect.n.01', 'anymore.r.01', '.', "''", 'john.n.03', 'suppose.v.01', '.'] 

EXEPTION ERROR: flew JJ
EXEPTION ERROR: nest JJS
Post_Lesked_Sentence:  ['group.n.03', '.'] 

EXEPTION ERROR: flew JJ
EXEPTION ERROR: nest JJS
Post_Lesked_Sentence:  ['together.r.05', '.'] 

Post_Lesked_Sentence:  ['womanhood.n.02', 'toy.v.02', 'violin.n.01', '.'] 

Post_Lesked_Sentence:  ['young.a.01', 'lady.n.03', 

In [8]:
# CAREFUL WITH THE STS.gs FILE. 
# NEEDS TO BE CORRECTED TO 54321 instead of 12345

for line in open('./trial/STS.gs.txt','r'):
    line = line.strip().split("\t")
    standard.append(int(line[1]))
    
a = pearsonr(standard[0:6], without_stop_words_tests[0:6])[0]
print('Pearson correlation with lesk:', round(a,3))

Pearson correlation with lesk: 0.36


Interestingly enough, the pearson coefficient is reduced by a great amount once we take away the stopwords